### ISSUE TITLE
TraceQL autocomplete

### ISSUE THREAD

**Issue by mapno** [grafana/tempo#1868]
## Context

As we develop TraceQL and its main integration with Grafana, it is important for the language that we make it easy to use and provide a gentle learning curve. A big part of that experience is making writing queries easy for the users.

Many well-known query languages, such as PromQL and LogQL, provide the user with suggestions of possible conditions that will result in a successful query. We aim to add this concept to TraceQL as well.

As it is right now, suggestions are very limited by the API design and implementation. Grafana queries Tempo for all tags and their values, and Tempo will return everything that it has, without any filtering based on the current query. That makes it possible for Grafana to suggest tags and values that produce no matches in Tempo.

## Objectives

- Improve tag values search results, so only values that can produce a query that returns a non-empty results set are returned.
- Design new API that makes querying for tags and values more efficient that current `# of queries = 1 tag query + (tag values queries * # of tags)`.

#### Objective 1: Extend `/api/v2/search/tag/<tag.name>/values`

- [x] Support query param that contains TraceQL query
  - eg. `/api/v2/search/tag/resource.namespace/values?q="{cluster="foo"}"`
- [x] Parse TraceQL query and extract conditions to filter
- [x] Implement filtering in WAL blocks
- [ ] Integrate in Grafana - https://github.com/grafana/grafana/issues/66956
- [x] Implement filtering in backend blocks
  - [x] Add non-collecting column iterators to tag value search
- [ ] ~Implement filtering in live traces (?)~ (Implemented using the traceql engine, which is not supported in live traces)
- [x] Compare performance between filtering / no filtering
- [ ] (Stretch) Support more than one spanset selector `{}` per query—ie. `{} && {}`.
  - [ ] ~Define identifier for the active spanset selector (eg. cursor position)~ No longer relevant

#### Objective 2: Improvements over Objective 1

- [ ] Support complete TraceQL spec
  - [ ] Support multiple spanset filters `{ true } && { true }`
  - [ ] Support OR conditions `{ true || false }`, `{ true } || { false }`
- [ ] Use yacc parser instead of regex
  - [ ] Implement graceful failure mode for parser

#### Objective 3 (Optional): Design and implementation of new API

Extending the current API has its limitations. Performance could become a significant bottleneck as the query that's being typed grows too long or complicated, since Tempo would be basically doing a search for every tag, fetching its values.

That's why, there's already been some consideration on how an improved solution could look like.

- [ ] Design new API together with Grafana
  - https://github.com/grafana/grafana/issues/60334
  - Current proposal based on `/series` API of [Prometheus](https://prometheus.io/docs/prometheus/latest/querying/api/#finding-series-by-label-matchers) and [Loki](https://grafana.com/docs/loki/latest/api/#list-series)

**Comment by joe-elliott** [comment]
I believe phase 1 and the first checkbox of phase 2 is complete. I'd also not worry about the in memory traces and only search the WAL for attributes since that is what search is doing anyway.

**Comment by 97amarnathk** [comment]
@joe-elliott @mapno I find this interesting and would like to work on this if this issue still open.

**Comment by mapno** [comment]
Hi @97amarnathk. Thanks for the interest, unfortunately, I'm already working on this issue. Apologies for not noting it here, my bad. Although I'm sure there'll be plenty of tasks that can be done around this project, it's still on its early stages. I will ping you in those.

**Comment by 09jvilla** [comment]
> (Stretch) Support more than one spanset selector {} per query

@mapno -- This means that with what you intend to deliver in the next 3 months, I'll be able to see the filtered tags if I've already specified `cluster=foo`, but if I have as a filter `cluster=foo` && `environment=dev`, you can't show me a filtered view that incorporates both of those?

**Comment by 09jvilla** [comment]
> only values that can produce a successful query are returned.

To be clear, when you say "successful" what you mean is "only values that can produce a query that returns a non-empty results set" right?

**Comment by mapno** [comment]
> Oh actually reading this more closely, what you're saying you won't support out of the gate is filtering for a query like `{cluster=foo} && {environment=dev}`. You would support `{cluster=foo && environment=dev}` Am I right?

That's exactly it

>To be clear, when you say "successful" what you mean is "only values that can produce a query that returns a non-empty results set" right?

Yes. That's a better way of saying it.

**Comment by 09jvilla** [comment]
Awesome - thanks for clarifying. 

> you won't support out of the gate is filtering for a query like {cluster=foo} && {environment=dev}. You would support {cluster=foo && environment=dev}

This seems like the right way to phase out the work. @adrapereira 's initial implementation of the visual query builder for traceQL only let's you create queries with 1 spanset, so what you're doing with objective 1 will cover what the frontend needs. =)

**Comment by 09jvilla** [comment]
Autocomplete bug found in the Grafana frontend:
https://github.com/grafana/tempo/issues/2200

Think this would be nice if we could add this fix to Objective 1 since `status` is a pretty common field people want to filter on.

**Comment by 09jvilla** [comment]
@mapno does objective 1 also include any changes to `GET /api/search/tags` (in other words, the endpoint that lists the tag names themselves)? 

As I understand it, what you have in objective 1 will handle filtering down to valid values (i.e. values that will produce a non-zero result). 

So if all my traces with tag `app=app1` also have the tag `dc` with values `dc1` and `dc2` but not `dc3` (which exists on traces with tag `app=app2`), then I would expect autocomplete to just return `dc1` and `dc2` (not `dc3`) when I pass the `app=app1` filter). 

However, imagine I have some spans with `app=app3`. None of those spans have the `dc` attribute. 

Will I be able to use autocomplete to tell me that? I think the answer is yes, but not via the `search tags` endpoint, but probably through `search tag values`? I'd have to hit `search tag values` with `app=app3` for the values of `dc`, see that there were no valid values for `dc`, and then conclude that `dc` isn't an attribute attached to spans with `app=app3`? 

I'm asking because I'm trying to imagine what this will look like in the Grafana frontend. I'm curious if we'll be able to filter down the list of tag names we let the user select (not just tag values) given their choices of `Service Name` and `Span Name` (or any other already applied key<>value filter).  

<img width="816" alt="no-options-found" src="https://user-images.githubusercontent.com/9610816/224869484-561dd6bc-120c-432b-a268-5fd2968ccf3d.png">

**Comment by mapno** [comment]
>Will I be able to use autocomplete to tell me that? I think the answer is yes, but not via the search tags endpoint, but probably through search tag values? I'd have to hit search tag values with app=app3 for the values of dc, see that there were no valid values for dc, and then conclude that dc isn't an attribute attached to spans with app=app3?

Exactly that. I wasn't planning on extending `/search/tags`, although I suppose that once implemented for tag values, it wouldn't be that difficult. The main drawback is that it would require a change on how Grafana interacts with the API—currently queries `/search/tags` once at the beginning. I'm not sure it's worth the effort.

**Comment by 09jvilla** [comment]
> The main drawback is that it would require a change on how Grafana interacts with the API—currently queries /search/tags once at the beginning. I'm not sure it's worth the effort.

Let's cross that bridge when we get there (specifically "let's cross that bridge when objective 1 is complete"). Perhaps not worth the effort and/or the performance would be crappy but if the alternative is waiting on objective 2 which is a multi-month effort, we may want to consider doing this as a stopgap if this is low effort. 

WDYT?

**Comment by 09jvilla** [comment]
I've created this bug issue so we don't lose this. However, as I mention above, I think fixing this bug is lower priority than getting to objective 1. 
https://github.com/grafana/tempo/issues/2202
